osf/migrations/0163_populate_file_versions_through.py

```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.15 on 2019-03-03 17:52
+from __future__ import unicode_literals
+
+import logging
+from math import ceil
+
+from django.db import migrations, connection
+
+logger = logging.getLogger(__file__)
+
+increment = 100000
+
+def restore_default_through_table(state, schema):
+    sql = """
+
+        INSERT INTO osf_basefilenode_versions (basefilenode_id, fileversion_id)
+        SELECT new_thru.basefilenode_id, new_thru.fileversion_id
+        FROM osf_basefileversionsthrough new_thru;
+    """
+    with connection.cursor() as cursor:
+        cursor.execute(sql)
+
+
+# Batching adapted from strategy in website/search_migration/migrate.py
+def populate_fileversion_name(state, schema):
+    BaseFileNodeVersion = state.get_model('osf.basefilenode_versions')
+    max_thru_id = getattr(BaseFileNodeVersion.objects.order_by('id').last(), 'id', 0)
+
+    sql = """
+        INSERT INTO osf_basefileversionsthrough (basefilenode_id, fileversion_id, version_name)
+        SELECT THRU.basefilenode_id, THRU.fileversion_id, F.name
+        FROM osf_basefilenode_versions THRU, osf_basefilenode F
+        WHERE THRU.basefilenode_id = F.id
+        AND THRU.id > {}
+        AND THRU.id <= {};
+    """
+
+    total_pages = int(ceil(max_thru_id / float(increment)))
+    page_start = 0
+    page_end = 0
+    page = 0
+    while page_end <= (max_thru_id):
+        page += 1
+        page_end += increment
+        if page <= total_pages:
+            logger.info('Transferring BaseFileNodeVersions to custom through table: {} / {}'.format(page_end / increment, total_pages))
+        with connection.cursor() as cursor:
+            cursor.execute(sql.format(
+                page_start,
+                page_end,
+                page_start,
+                page_end
+            ))
+        print page
+        page_start = page_end
```

Comment:
This is the latest version of this...create table as select is fast, but it doesn't have any constraints, indices, types...so we have to add that back in. But then there's the issue of linking this Django model to this table.  
```    
CREATE TABLE osf_base_filenode_versions_through AS
    SELECT
        obfv.basefilenode_id,
        obfv.fileversion_id,
        ob.name
    FROM
        osf_basefilenode_versions obfv
        LEFT JOIN
            osf_basefilenode ob
            ON obfv.basefilenode_id = ob.id;
    ALTER TABLE osf_base_filenode_versions_through ADD COLUMN id SERIAL PRIMARY KEY;
    ALTER TABLE osf_base_filenode_versions_through ADD CONSTRAINT osf_basefilenod_basefilenode_id_b0nwad27_fk_osf_basefilenode_id FOREIGN KEY (basefilenode_id) REFERENCES osf_basefilenode DEFERRABLE INITIALLY DEFERRED;
    ALTER TABLE osf_base_filenode_versions_through ALTER COLUMN basefilenode_id
    SET
        DATA TYPE INTEGER;
    ALTER TABLE osf_base_filenode_versions_through ALTER COLUMN fileversion_id
    SET
        NOT NULL;
    ALTER TABLE osf_base_filenode_versions_through ALTER COLUMN fileversion_id
    SET
        DATA TYPE INTEGER;
    ALTER TABLE osf_base_filenode_versions_through ALTER COLUMN basefilenode_id
    SET
        NOT NULL;
    ALTER TABLE osf_base_filenode_versions_through ADD CONSTRAINT osf_basefilenode__fileversion_id_93nwadfc_fk_osf_fileversion_id FOREIGN KEY (fileversion_id) REFERENCES osf_fileversion DEFERRABLE INITIALLY DEFERRED;
    ALTER TABLE osf_base_filenode_versions_through ADD CONSTRAINT dist_id_zipcode_key UNIQUE (basefilenode_id, fileversion_id);
    CREATE INDEX
    ON osf_base_filenode_versions_through (basefilenode_id, fileversion_id);
    CREATE INDEX
    ON osf_base_filenode_versions_through (basefilenode_id);
    CREATE INDEX
    ON osf_base_filenode_versions_through (fileversion_id);
```

Reply:
So it looks like we're around 7 minutes locally to get it moved and indexed? Hopefully just naming the table properly will do most of the rest of the work for the linking to django. I had put a lot of extra underscores in the create to keep from interfering with whatever you had.

Reply:
I haven't yet figured out how to get the Django model connected to this table, it's tricky!

Reply:
So `drop`ping the table that django makes and replacing it with the above (named properly) doesn't do it?

Reply:
I'm probably not naming it properly - i did `python manage.py inspectdb` and tried to take the name that it output and use that, and variations of it, I tried defining `db_table` under Meta - changes that were having me to make more migrations.  still haven't figured out how to link them, and I don't know know if it's a simple as their names matching, if you do it backwards like this.  hopefully it is.

Reply:
ORMs are super convenient right up until when they aren't. :)